Google.Api.Generator/Generation/ResourceNamesGenerator.cs

```diff
@@ -524,23 +524,41 @@ private IEnumerable<ClassDeclarationSyntax> ProtoMessagePartials()
         {
             // Note: Whether a field is required or optional is purposefully ignored in this partial class.
             // The optionalness of a field is only relevant within a method signature (flattening).
-            foreach (var msg in _fileDesc.MessageTypes)
+            foreach (var topLevelMessage in _fileDesc.MessageTypes)
             {
-                var resources = msg.Fields.InFieldNumberOrder()
-                    .Select(fieldDesc => (fieldDesc, resDetails: _catalog.GetResourceDetailsByField(fieldDesc)))
-                    .Where(x => x.resDetails != null)
+                // First find all resource-referencing fields for all messages recursively.
+                // We then pick only the relevant ones to generate on a per-message basis.
+                var allResources = topLevelMessage.SelfAndNestedMessagesRecursively()
+                    .SelectMany(m => m.Fields.InFieldNumberOrder())
+                    .Select(_catalog.GetResourceDetailsByField)
+                    .Where(x => x is not null)
+                    .SelectMany(r => r)
                     .ToList();
-                if (resources.Any())
+
+                var messagesContainingResources = allResources
+                    .Select(r => r.Descriptor.ContainingType)
+                    .SelectMany(m => m.SelfAndAncestors())
+                    .ToHashSet();
+
+                if (allResources.Any())
+                {
+                    yield return GenerateClassWithResourceReferences(topLevelMessage);
+                }
+
+                ClassDeclarationSyntax GenerateClassWithResourceReferences(MessageDescriptor msg)
```

Comment:
Is this capturing anything from within the foreach scope? If not, maybe declare it outside the foreach, even if for dropping the extra indent.

Reply:
Will give that a try.

Reply:
It captures `allResources` and `messagesContainingResources` which are both declared in the loop. (We could potentially make them file-descriptor wide, but I don't think it's worth it.) Merging as-is.